pkg/resource/deploy/step_generator.go

```diff
@@ -611,9 +611,14 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, res
 		return []Step{NewImportStep(sg.deployment, event, new, goal.IgnoreChanges, randomSeed)}, nil
 	}
 
+	isTargeted := sg.isTargetedForUpdate(new)
+	if isTargeted {
+		sg.updateTargetsOpt.addLiteral(urn)
+	}
+
 	// Ensure the provider is okay with this resource and fetch the inputs to pass to subsequent methods.
 	var err error
-	if prov != nil {
+	if isTargeted && prov != nil {
```

Comment:
We've made a very similar change to this before (https://github.com/pulumi/pulumi/pull/9960) which triggered a pretty bad P0 (https://github.com/pulumi/pulumi/issues/10020).

I'd be pretty hesitant to make changes like this again unless we think we have tests to explain and cover what happened back then (https://github.com/pulumi/pulumi/issues/10056).

Reply:
I'm looking at the code you linked, and it had some early exits. This change makes the branch more selective and doesn't call prov.Check() instead.

I'll add a test to add coverage for this issue https://github.com/pulumi/pulumi/issues/10056

Reply:
The issue the user ran into was very vague and I cannot reproduce it in a reasonable amount of time.